src/jit/morph.cpp

```diff
@@ -83,6 +83,18 @@ GenTreePtr          Compiler::fgMorphIntoHelperCall(GenTreePtr      tree,
     tree->gtCall.gtEntryPoint.addr = nullptr;
 #endif
 
+#if defined(_TARGET_X86_) && !defined(LEGACY_BACKEND)
+    if (varTypeIsLong(tree))
+    {
+        GenTreeCall* callNode = tree->AsCall();
+        ReturnTypeDesc* retTypeDesc = callNode->GetReturnTypeDesc();
+        retTypeDesc->Reset();
+        retTypeDesc->Initialize(this, callNode->gtRetClsHnd);
+        
+        NYI("Helper with TYP_LONG return type");
+    }
+#endif
+
```

Comment:
I think I ran into this issue myself. As far as I understand `TYP_LONG` call arguments aren't decomposed so a decomposed call tree can still contain `GT_LONG` nodes. This works fine until you run into a call that is itself `TYP_LONG`. Now the call tree is moved to an embedded statement and that statement will be decomposed after the current (top-level) statement is decomposed. But the tree of that embedded statement was already decomposed...

In short, we need `GTF_STMT_SKIP_DECOMPOSE` or something similar. Though I wonder why ignoring `GT_LONG` nodes in decomposition didn't work.